### PR TITLE
MFT: first translation from AliRoot of the standalone tracking code

### PR DESCRIPTION
**Pull request by bovulpes**
This is the first version of the standalone tracking code adapted from the AliRoot MFT code. It implements the two methods of track finding: the linear-track-finder (LTF) and the one based on a cellular automaton (CA), applied in this order.
It introduces the sorting of the clusters in R:Phi bins, in order to speed up the cluster combinations, together with the search within an R:Phi bins window, in order to reduce the number of possible cluster combinations. 
This last feature makes the algorithm independent of the presence of an external vertex information, while covering the displacement of the vertex in z.
The result is expressed in two streamable track objects, TrackLTF and TrackCA. A special writer/reader will come with a future PR, together with a common track model, including both mentioned.

### CODE REVIEW

File: Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/ClusterReaderSpec.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// @file   DigitReaderSpec.h
```
Comment (sawenzel) [inline Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/ClusterReaderSpec.h]: probably a copy-paste-error

File: Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/ClusterReaderSpec.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+#ifndef O2_MFT_CLUSTERREADER_H_
+#define O2_MFT_CLUSTERREADER_H_
+
+#include "TFile.h"
+
+#include "Framework/DataProcessorSpec.h"
+#include "Framework/Task.h"
+
+using namespace o2::framework;
```
Comment (sawenzel) [inline Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/ClusterReaderSpec.h]: we should avoid using `using namespace` in header files

**Comment by bovulpes** [comment]
Ok, thanks for the suggestions.

**Comment by sawenzel** [comment]
@dberzano : It seems like the python scripts of the codechecker are no longer running: 
```
ThinCompilationsDatabase.py -exclude-files '(?:.*G\_\_.*\.cxx|.*\.pb.cc)' -use-files Detectors/ITSMFT/MFT/base/src/GeometryTGeo.cxx:Detectors/ITSMFT/MFT/reconstruction/include/MFTReconstruction/ClustererTask.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Cell.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Cluster.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Constants.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Definitions.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/IOUtils.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/IndexTableUtils.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/MathUtils.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/ROframe.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Road.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h:Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Tracker.h:Detectors/ITSMFT/MFT/tracking/src/Cluster.cxx:Detectors/ITSMFT/MFT/tracking/src/IOUtils.cxx:Detectors/ITSMFT/MFT/tracking/src/MFTTrackingLinkDef.h:Detectors/ITSMFT/MFT/tracking/src/ROframe.cxx:Detectors/ITSMFT/MFT/tracking/src/Tracker.cxx:Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/ClusterReaderSpec.h:Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/ClusterWriterSpec.h:Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/ClustererSpec.h:Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/DigitReaderSpec.h:Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/RecoWorkflow.h:Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/TrackWriterSpec.h:Detectors/ITSMFT/MFT/workflow/include/MFTWorkflow/TrackerSpec.h:Detectors/ITSMFT/MFT/workflow/src/ClusterReaderSpec.cxx:Detectors/ITSMFT/MFT/workflow/src/ClusterWriterSpec.cxx:Detectors/ITSMFT/MFT/workflow/src/ClustererSpec.cxx:Detectors/ITSMFT/MFT/workflow/src/DigitReaderSpec.cxx:Detectors/ITSMFT/MFT/workflow/src/RecoWorkflow.cxx:Detectors/ITSMFT/MFT/workflow/src/TrackWriterSpec.cxx:Detectors/ITSMFT/MFT/workflow/src/TrackerSpec.cxx:Detectors/ITSMFT/MFT/workflow/src/mft-reco-workflow.cxx
Traceback (most recent call last):
  File "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/Python-modules/1.0-13/lib/python/site-packages/site.py", line 73, in <module>
    __boot()
  File "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/Python-modules/1.0-13/lib/python/site-packages/site.py", line 3, in __boot
    import os
ImportError: No module named os
```

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Constants.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+{
+
+namespace Constants
+{
+
+namespace Math
+{
+constexpr Float_t Pi{ 3.14159265359f };
+constexpr Float_t TwoPi{ 2.0f * Pi };
+constexpr Float_t FloatMinThreshold{ 1e-20f };
```
[the remark this answers is not in the document]
Comment (bovulpes) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Constants.h]: You're right, I will use as much as I can.

**Comment by sawenzel** [comment]
This PR now fails with: `#include "AliTPCCommonDefGPU.h" not found`. @davidrohr : Is this due to the move of AliTPCCommon? How should the code here be adapted?

**Comment by bovulpes** [comment]
Could you please let this PR merge, since we have not enough men power and the way it goes it will never be fine with the rest of O2. I will continue to work on it and implement the suggestions from Ruben.

**Comment by sawenzel** [comment]
Sorry for the disruptions / merge conflicts.

**Comment by davidrohr** [comment]
After the merge of AliTPCCommon, you have to use GPUCommonDef.h instead of AliTPCCommonDefGPU.h, perhaps you did not rebase correctly?

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/ROframe.h [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+  Bool_t isClusterUsed(Int_t layer, Int_t clusterId) const;
+  void markUsedCluster(Int_t layer, Int_t clusterId);
+
+  void clear();
+
+ private:
+  Int_t mROframeId;
+  std::array<std::vector<Cluster>, Constants::MFT::LayersNumber> mClusters;
+  std::array<std::vector<MCCompLabel>, Constants::MFT::LayersNumber> mClusterLabels;
```
[the remark this answers is not in the document]
Comment (bovulpes) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/ROframe.h]: I will do it in a future commit.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/ROframe.h [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+  return mCells[layerId];
+}
+
+template <typename... T>
+void ROframe::addClusterToLayer(Int_t layer, T&&... values)
+{
+  mClusters[layer].emplace_back(std::forward<T>(values)...);
+}
+
+inline void ROframe::addClusterLabelToLayer(Int_t layer, const MCCompLabel label) { mClusterLabels[layer].emplace_back(label); }
```
[the remark this answers is not in the document]
Comment (bovulpes) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/ROframe.h]: I leave it for a future commit.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+{
+  mCellLayer.push_back(layer);
+  mCellId.push_back(cellId);
+}
+
+inline void TrackCA::removeLastCell(Int_t& layer, Int_t& cellId)
+{
+  layer = mCellLayer.back();
+  cellId = mCellId.back();
+
```
[the remark this answers is not in the document]
Comment (bovulpes) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h]: The tracker uses only the interface function Tracker::addCellToCurrentTrackCA
which ensures that all the other vectors of a track are updated every time a
new cell is added, so addCell and setPoint work in tandem on all vectors of
the track.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+  const std::vector<Float_t>& getYCoordinates() const { return mY; }
+  const std::vector<Float_t>& getZCoordinates() const { return mZ; }
+  const std::vector<Int_t>& getLayers() const { return mLayer; }
+  const std::vector<Int_t>& getClustersId() const { return mClusterId; }
+  const std::vector<Int_t>& getCellsLayer() const { return mCellLayer; }
+  const std::vector<Int_t>& getCellsId() const { return mCellId; }
+  const std::vector<MCCompLabel>& getMCCompLabels() const { return mMCCompLabels; }
+
+ private:
+  Int_t mRoadId{ -1 };
```
[the remark this answers is not in the document]
Comment (bovulpes) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h]: Could you please explain?

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+  Float_t mChiSquareZX{ 0. };
+  Float_t mChiSquareZY{ 0. };
+  std::vector<Float_t> mX;
+  std::vector<Float_t> mY;
+  std::vector<Float_t> mZ;
+  std::vector<Int_t> mLayer;
+  std::vector<Int_t> mClusterId;
+  std::vector<Int_t> mCellLayer;
+  std::vector<Int_t> mCellId;
+  std::vector<MCCompLabel> mMCCompLabels;
```
[the remark this answers is not in the document]
Comment (bovulpes) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h]: I allowed myself a certain redundance at the beginning, this info will be
reduced to a minimum in some future.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+  const Int_t getNPoints() const { return mX.size(); }
+  const std::vector<Float_t>& getXCoordinates() const { return mX; }
+  const std::vector<Float_t>& getYCoordinates() const { return mY; }
+  const std::vector<Float_t>& getZCoordinates() const { return mZ; }
+  const std::vector<Int_t>& getLayers() const { return mLayer; }
+  const std::vector<Int_t>& getClustersId() const { return mClusterId; }
+  const std::vector<MCCompLabel>& getMCCompLabels() const { return mMCCompLabels; }
+
+ private:
+  std::vector<Float_t> mX;
```
[the remark this answers is not in the document]
Comment (bovulpes) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h]: Could you explain?

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Tracker.h [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+  Bool_t getBinClusterRange(const ROframe&, const Int_t, const Int_t, Int_t&, Int_t&) const;
+  const Float_t getCellDeviation(const ROframe&, const Cell&, const Cell&) const;
+  const Bool_t getCellsConnect(const ROframe&, const Cell&, const Cell&) const;
+  const Float_t getCellChisquare(ROframe&, const Cell&) const;
+  const Bool_t addCellToCurrentTrackCA(const Int_t, const Int_t, ROframe&);
+
+  const Bool_t LinearRegression(Int_t, Float_t*, Float_t*, Float_t*, Float_t&, Float_t&, Float_t&, Float_t&, Float_t&, Int_t skip = -1) const;
+
+  Float_t mBz = 5.f;
+  std::uint32_t mROFrame = 0;
```
[the remark this answers is not in the document]
Comment (bovulpes) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Tracker.h]: Yes, todo.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Cell.h [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+ private:
+  const Int_t mFirstLayerId;
+  const Int_t mSecondLayerId;
+  const Int_t mFirstClusterIndex;
+  const Int_t mSecondClusterIndex;
+  Int_t mLevel;
+  Bool_t mUpdateLevel;
+  Bool_t mIsUsed;
+  Int_t mCellId;
+  std::vector<std::pair<Int_t, Int_t>> mLeftNeighbours;
```
[the remark this answers is not in the document]
Comment (bovulpes) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/Cell.h]: I tried another method and replaced the dynamical vectors by fixed arrays.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (shahor02) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h]: I meant why do you need to keep the coordinates in the track. Is this class internal to algorithm or they are stored / shipped to other device?

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (shahor02) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h]: ok, still, I would avoid having dynamical objects as data members of the track.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (shahor02) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackCA.h]: was again about using vectors

**Pull request closed without merging**